rosidl_cmake/cmake/rosidl_generate_interfaces.cmake

```diff
@@ -137,21 +137,48 @@ macro(rosidl_generate_interfaces target)
   # afterwards all remaining interface files are .idl files
   list(APPEND _idl_tuples ${_idl_adapter_tuples})
 
-  # to generate action interfaces, we need to depend on "action_msgs"
+  # Check for any action or service interfaces
+  # Which have implicit dependencies that need to be found
   foreach(_tuple ${_interface_tuples})
+    # We use the parent directory name to identify if the interface is an action or service
     string(REGEX REPLACE ".*:([^:]*)$" "\\1" _tuple_file "${_tuple}")
     get_filename_component(_parent_dir "${_tuple_file}" DIRECTORY)
+
     if("${_parent_dir}" STREQUAL "action")
+      # Actions depend on the packages service_msgs and action_msgs
+      find_package(service_msgs QUIET)
+      if(NOT ${service_msgs_FOUND})
+        message(FATAL_ERROR
+          "Unable to generate action interface for '${_tuple_file}'. "
+          "In order to generate action interfaces you must add a depend tag "
+          "for 'service_msgs' in your package.xml.")
+      endif()
+      ament_export_dependencies(service_msgs)
+      list_append_unique(_ARG_DEPENDENCIES "service_msgs")
       find_package(action_msgs QUIET)
       if(NOT ${action_msgs_FOUND})
         message(FATAL_ERROR
           "Unable to generate action interface for '${_tuple_file}'. "
           "In order to generate action interfaces you must add a depend tag "
           "for 'action_msgs' in your package.xml.")
       endif()
-      list_append_unique(_ARG_DEPENDENCIES "action_msgs")
       ament_export_dependencies(action_msgs)
+      list_append_unique(_ARG_DEPENDENCIES "action_msgs")
+
+      # It is safe to break out of the loop since services only depend on service_msgs
+      # Which has already been found above
       break()
+    elseif(NOT service_msgs_FOUND AND "${_parent_dir}" STREQUAL "srv")
```

Comment:
Ah, this is a bug. If the caller manually calls `find_package(service_msgs)` _before_ calling `rosidl_generate_interfaces`, then `service_msgs` will not be added to the list to the dependency list for linking against the target.

See [3168e55](https://github.com/ros2/rosidl/pull/700/commits/3168e55019442027a2dc99df93ac8c11a0e43486)
I just removed the use of `service_msgs_FOUND`. It's less efficient, but I don't think it should impact performance too badly.